### PR TITLE
Take non-`Send` resources out of `World`

### PR DESCRIPTION
**Pull request by maniwani**
This PR builds on top of #9202 (I separated what could be separated from this PR into that one).

## Objective

The aim of this PR is to be the "final" fix to our messy problems involving thread-local data (see https://github.com/bevyengine/bevy/discussions/6552 for context).

Bevy only deals with a handful of thread-local resources (that mostly come from low-level OS APIs or libs that talk to those), but that handful has resulted in a ton of internal complexity (in `bevy_ecs`, `bevy_tasks`, and `bevy_app`). The way we currently store them makes it either difficult or impossible to implement some nice-to-have features (e.g. generalized multi-world apps, framerate-independent input handling, etc.).

## Solution

**tl;dr** Separate non-`Send` data from the ECS entirely. It will remain indirectly accessible through an event loop running on the main thread. Then, move app updates into a separate thread (on permitting targets) so they don't block the event loop.

This setup should be familiar to any audio plugin developers (or `async` runtime developers). If a system on thread `B` needs to do something with data that can't leave thread `A`, it can just send a command to `A`'s event loop and wait for `A` to run it (and send any result back).

Some unique benefits of separating `Send` and `!Send` data like this are:
- `Res` and `ResMut` can be removed in favor of `Ref` and `Mut` since they're redundant without `NonSend` and `NonSendMut`.
- `World` becomes unambiguously `Send`, so a world can be dropped on any thread.
- The multi-threaded executor is able to run any system on any thread (including exclusive systems). 
- The multi-threaded executor is able to run any system without spawning a task for it (especially exclusive systems).
- Any system from any world on any thread can remotely access the local resources in the main thread.
- Input events could be timestamped (with `Instant::now()`) with virtually no delay/aliasing (the event loop isn't blocked).

---

## Changelog

TODO: finish this section

- `NonSend` and `NonSendMut` have been removed.
- Moved storage of non-`Send` resources to an external `ThreadLocals` container. It's normally owned by the `App`.

## Migration Guide

If you were doing this before:
```rust
fn foo(x: NonSend<X>) {
    /* ... */
}
```

you're doing this now:
```rust
fn foo(mut thread: ThreadLocal) {
    // This closure will run on the thread that owns the TLS.
    // `run` blocks until the closure is completed.
    thread.run(|tls| {
        let x = tls.resource::<X>();
        /* ... */
    });
}
```

For better or worse, the handling that used to be buried inside the scheduler is now the responsibility of whoever writes an `App` runner. It's not *too* much boilerplate though IMO. Here's an example of a basic runner that runs the app once.

```rust
fn run_once(mut app: App) {
    let (mut sub_apps, tls, send, recv, _runner) = app.into_parts();
          
    // Move sub-apps to another thread.
    let thread = std::thread::spawn(move || {
        let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
            sub_apps.update();
            send.send(AppEvent::Exit(sub_apps)).unwrap();
        }));
        
        if let Some(payload) = result.err() {
            send.send(AppEvent::Error(payload)).unwrap();
        }
    });
    
    // Run an event loop in this thread.
    loop {
        let event = recv.recv().unwrap();
        match event {
            AppEvent::RunTask(task) => {
                task();
            }
            AppEvent::Exit(_) => {
                thread.join().unwrap();
                break;
            }
            AppEvent::Error(payload) => {
                std::panic::resume_unwind(payload);
            }
        }
    }
}
```

Spawning a thread like this is basically mandatory unless you're targeting an environment (i.e. `wasm32`) or are enabling a `bevy` compiler flag that forces completely single-threaded system execution. Otherwise, not doing so will risk running into a deadlock.

~~Unfortunately, the channel between the main and app threads can't always be a `std::sync::mpsc::channel`. The `winit` runner spawns an event loop that can only be woken up by `winit` events, so we have to use their special `EventLoopProxy` sender. That's the reason for the `ThreadLocalTaskSender` trait.~~ I still had to box the channel as a trait object because `bevy_ecs` and `bevy_app` are separated.

### CODE REVIEW

**Review by hymm — COMMENTED** [review]
This is just a quick skim with comments on trying to shrink this pr a bit. Will dig more into the meat of the pr later.

File: crates/bevy_tasks/src/task_pool.rs [hunk 1]
```diff
@@ -373,140 +307,80 @@ impl TaskPool {
...
-                                Ok(res) => results.push(res),
-                                Err(payload) => std::panic::resume_unwind(payload),
-                            }
-                        } else {
-                            panic!("Failed to catch panic!");
-                        }
-                    }
-                    results
-                };
+                    let mut pending_tasks = Vec::with_capacity(spawned.len());
```
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_tasks/src/task_pool.rs:316]: Yeah, probably.

File: crates/bevy_app/src/headless.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
+use crate::{
```
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_app/src/headless.rs:1]: Yep. It's just the `ScheduleRunnerPlugin`. I forgot I had renamed it.

File: crates/bevy_app/src/sub_app.rs [hunk 3]
```diff
@@ -0,0 +1,408 @@
+use crate::{
```
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_app/src/sub_app.rs:1]: No.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 4]
```diff
@@ -525,122 +498,70 @@ impl MultiThreadedExecutor {
...
-            self.local_thread_running = true;
-            scope.spawn_on_external(task);
-        }
+        scope.spawn(task);
     }
 
     /// # Safety
     /// Caller must ensure no systems are currently borrowed.
-    unsafe fn spawn_exclusive_system_task<'scope>(
+    unsafe fn run_exclusive_system<'scope>(
```
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:506]: Sure.

File: crates/bevy_tasks/src/task_pool.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (hymm) [reply crates/bevy_tasks/src/task_pool.rs:316]: It sort of looks like an attempt to remove the assert unwind safe from the multi threaded executor by making scope use a fair polling Algo

[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_tasks/src/task_pool.rs:316]: Yeah, I based this branch on my other fair polling PR. I don't think changing `bevy_tasks` is required for this PR.

File: crates/bevy_ecs/src/storage/resource_non_send.rs [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+                &mut state.component_id,
+                system_meta,
+                world,
+                curr_tick,
+            );
+
+            let access = if std::thread::current().id() == accessor.owning_thread {
+                // SAFETY: We are on the owning thread and pointer is valid.
+                let ptr = accessor.direct;
+                unsafe { ThreadLocalAccess::Direct(&mut *ptr) }
```
Comment (SkiFire13) [inline crates/bevy_ecs/src/storage/resource_non_send.rs:201]: This is unsound since `ThreadLocalAccess` wasn't created with a mutable reference.

Comment (SkiFire13) [inline crates/bevy_ecs/src/storage/resource_non_send.rs:201]: Also, are we sure there's no other direct access to the thread local? For example, couldn't we access it through the `ThreadLocalAccess` that exist in a `SubApp`? In that case we would have two mutable references to the same shared thread local storage, which would be very bad.

**Review by SkiFire13 — COMMENTED** [review]
> ```rust
>     // Insert channel resource into every sub-app.
>     let (send, recv) = std::sync::mpsc::channel();
>     sub_apps.main.world.insert_resource(ThreadLocalAccessor::new(&mut tls, send.clone());
>     for sub_app in sub_apps.sub_apps.values_mut() {
>         sub_app.world.insert_resource(ThreadLocalAccessor::new(&mut tls, send.clone());
>     }
> ```

FYI this example is unsound, each `&mut tls` technically invalidate any pointer obtained from the previous `&mut tls`.

File: crates/bevy_ecs/src/storage/resource_non_send.rs (hunk 5 above)
Comment (maniwani) [reply crates/bevy_ecs/src/storage/resource_non_send.rs:201]: For now, I'll change it to an `unsafe` method that requires passing in a `*mut ThreadLocals` directly.

Comment (maniwani) [reply crates/bevy_ecs/src/storage/resource_non_send.rs:201]: No, unless it's somehow possible for two systems to run on the same thread at the same time. There's also no public API on `ThreadLocalAccessor` that lets you read the pointer.

Comment (maniwani) [reply crates/bevy_ecs/src/storage/resource_non_send.rs:201]: Is it? I'm not familiar with what happens to references when a future is awaited and yields to another, but `ThreadLocals` is non-`Send` and non-`Sync`, so it can only be referenced on the same thread, so AFAIK it's impossible to see aliasing.

Comment (SkiFire13) [reply crates/bevy_ecs/src/storage/resource_non_send.rs:201]: I was thinking of something like this:

```rs
fn bad_subapp_extract(main_world: &mut World, subapp: &mut App) {
    let main_tl_state = SystemState::<ThreadLocal>::new(main_world);
    let main_tl = main_tl_state.get_mut(main_world);

    // This may run on this thread or another thread, doesn't really matter.
    main_tl.run(|tls1| {
        // Here we're in the tls thread, so this will get direct access to the tls
        let subapp_tl_state = SystemState::<ThreadLocal>::new(&mut subapp.world);
        let subapp_tl = subapp_tl_state.get_mut(&mut subapp.world);

        // Thus this should run immediately in this thread
        subapp_tl.run(|tls2| {
            // Now we have access to both tls1 and tls2, bad!
        });
    });
}
```

Where would this fail?

Comment (maniwani) [reply crates/bevy_ecs/src/storage/resource_non_send.rs:204]: :/ Well, I stand corrected.

Could either:
- Get rid of the direct access and only have it when the target is `wasm32` (or some compiler flag is enabled).
- Work a reference counter in there somehow and panic if there is more than one reference.

Comment (maniwani) [reply crates/bevy_ecs/src/storage/resource_non_send.rs]: OK, the current solution I've come up with is to use a wrapped `Box<Mutex<ThreadLocals>` so that the pointer stays valid and every borrow has to acquire the lock to create the `&mut ThreadLocals`.

File: crates/bevy_app/src/sub_app.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_app/src/sub_app.rs:1]: I totally missed the word "precursor." Yeah, I think that can be split out and merged first.

File: crates/bevy_ecs/src/storage/resource_non_send.rs [hunk 6]
```diff
@@ -0,0 +1,595 @@
...
+
+// SAFETY: The pointer to the thread-local storage is only dereferenced in its owning thread.
+unsafe impl Send for ThreadLocalChannel {}
+
+// SAFETY: The pointer to the thread-local storage is only dereferenced in its owning thread.
+// Likewise, all operations require an exclusive reference, so there can be no races.
+unsafe impl Sync for ThreadLocalChannel {}
+
+/// Guards access to its thread-local data storage.
+pub struct ThreadLocalStorage {
```
Comment (hymm) [inline crates/bevy_ecs/src/storage/resource_non_send.rs:389]: what's the advantage of the approach here vs building an abstraction around a global `thread_local!()`?

File: crates/bevy_app/src/sub_app.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_app/src/sub_app.rs:1]: Ok, I made #9202.

**Comment by alice-i-cecile** [comment]
I think we should swap this to the 0.13 milestone now. We're drawing close and this is still in draft.

**Comment by maniwani** [comment]
Let me see if I can have this ready for review this weekend. I've already rebased (just haven't pushed) and only the `winit` runner is unfinished.

**Comment by maniwani** [comment]
I will move this out of draft soon. There is at least one thing to debug still. (The game examples crash after a couple seconds.)

Decided to push this to the next milestone after all since it'll need some thorough testing. It'll also be easier to review after #8745 and #9202 are merged. I'd still like to see this merged as soon as possible since it's blocking other important things.

File: crates/bevy_ecs/src/storage/resource_non_send.rs [hunk 7]
```diff
@@ -0,0 +1,605 @@
...
+            let mutex = unsafe { &*ptr };
+            let guard = mutex
+                .try_lock()
+                .expect("lock on `ThreadLocals` should be available");
+
+            ThreadLocalAccess::Direct(ThreadLocalsGuard { guard, _arc: arc })
+        } else {
+            let ptr: *mut dyn ThreadLocalTaskSender = &mut *accessor.indirect;
+            // SAFETY: The pointer is valid. We have to do this to satisfy the borrow checker.
+            ThreadLocalAccess::Indirect(unsafe { &mut *ptr })
```
Comment (SkiFire13) [inline crates/bevy_ecs/src/storage/resource_non_send.rs:597]: Couldn't this be `ThreadLocalAccess::Indirect(&mut *accessor.into_inner().indirect)`?

Comment (maniwani) [reply crates/bevy_ecs/src/storage/resource_non_send.rs:597]: Thanks! Nice catch.

**Comment by maniwani** [comment]
Leaving a note to myself. #9826 made it possible for the render thread to spawn before the runner is called, so we can't add the `ThreadLocalChannel` resource to the `RenderApp` in `winit_runner`. We need `ThreadLocalChannel` to implement `Clone` so the main world can add it to the render world before/during extraction.

**Comment by lee-orr** [comment]
Couldn't you spawn a thread wrapping the `EventLoopProxy` within the winit plugin, and then send messages through a normal channel from there?

**Comment by maniwani** [comment]
> Could spawning a thread wrapping the `EventLoopProxy` within the `winit` plugin, and then sending messages through a normal channel from there work?

Are you suggesting a something like this?

```rust
// create a standard rust channel
let (send, recv) = std::sync::mpsc::channel();

// create a winit proxy
let proxy = event_loop.create_proxy();

// move it into a new thread (so there would be thread "main", thread "app", and now this)
std::thread::Builder::new()
    .name("winit-proxy-listener".into())
    .spawn(move ||
        while let Ok(event) = recv.recv() {
            // got something from the app, forward it to the event loop
            proxy.send(event).unwrap();
        }
    )
    .unwrap();
```

And were specifically addressing my last comment or what I wrote in the description about the `ThreadLocalTaskSender`? (Or both?) It does seem like an option to address both problems at once, so I do like your idea. It feels a little overkill to create a thread just to forward messages though. I wonder if the overhead from just having threads is negligible.

**Comment by maniwani** [comment]
Also, side note, `ThreadLocalChannel` should implement `Clone` anyway. Once this PR is merged, I want to reorder the app construction process so that building plugins is deferred[^1] and only happens after the runner starts, so the channel will already exist when the `RenderApp` is created. We'd clone it and add it to the render world then.

[^1]: The only thing a plugin would be able to do immediately on import is change the runner.

**Comment by lee-orr** [comment]
That was roughly what I was thinking, yes.

It was a reply to both, plus a question since I'm dealing with some thread-related things in the context of hot reload and was curious if there was a reason it wasn't possible...

Although given that concern - maybe it could be a specialized thing that happens in the app thread as a task? The main thought is wrapping it up somehow to allow the communication into the world to rely purely on standard channels.

**Comment by lee-orr** [comment]
> Also, side note, `ThreadLocalChannel` should implement `Clone` anyway. Once this PR is merged, I want to reorder the app construction process so that building plugins is deferred[1](#user-content-fn-1-f502f6a44712d4e3447a7d1e560d88b2) and only happens after the runner starts, so the channel will already exist when the `RenderApp` is created. We'd clone it and add it to the render world then.
> ## Footnotes
> 
>     1. The only thing a plugin would be able to do immediately on import is change the runner. [↩](#user-content-fnref-1-f502f6a44712d4e3447a7d1e560d88b2)

This actually brings up a question I've been mulling over - whether it's worth having two distinct plugin types: "Frame" plugins (that require non-send/sync elements and can't be disposed and re-created easily, like winit) and normal plugins (where either it's fully send/sync, or the elements that are non-send/sync are just optimizations/caching/etc and can be thrown away and re-created as needed - for example having a thread-local RNG instance)

I'm coming at this from the perspective of working on hot-reload, where having a fully separate "frame" would allow for much easier reload of the internals.

**Comment by maniwani** [comment]
> The main thought is wrapping it up somehow to allow the communication into the world to rely purely on standard channels.

Yeah, I think I'll try incorporating this idea. I am pretty unsatisfied with the `ThreadLocalTaskSender` trait, especially since it's just a byproduct of trying to interface with `winit`. It makes sense for `WinitPlugin` to adopt that complexity.

**Comment by lee-orr** [comment]
Is there an existing discussion related to plugin initialization and stuff? Somewhere where the "frame" idea either was discussed or could be raised effectively?

**Comment by maniwani** [comment]
> This actually brings up a question I've been mulling over - whether it's worth having two distinct plugin types: "Frame" plugins (that require non-send/sync elements and can't be disposed and re-created easily, like winit) and normal plugins (where either it's fully send/sync, or the elements that are non-send/sync are just optimizations/caching/etc and can be thrown away and re-created as needed - for example having a thread-local RNG instance).

I don't think we need this. The only thing in a plugin that "can't wait" is setting the runner function. The `WinitPlugin` only creates the `EventLoop` during its `build` method because the `RenderPlugin` needs a window to initialize the renderer on some targets.

If we defer plugins being built, then the `EventLoop` can just be created by the runner function. It won't need to be stored in TLS.

> Is there an existing discussion related to plugin initialization and stuff? Somewhere where the "frame" idea either was discussed or could be raised effectively? (I don't want to pollute this PR with that)

I don't know of anything recent, so free to start one.

**Comment by lee-orr** [comment]
Thanks for the quick reply! I agree that it's not needed for this PR to work well - I think it's something that might be useful for other elements (like the editor, embedding in non-bevy apps, and hot reload).  I'll create a new discussion for that, but regardless am excited for this PR!

**Comment by alice-i-cecile** [comment]
> If we defer plugins being built

For context, I really want to do this anyways to address #1255.
